### PR TITLE
add possibility to put Patchouli-Books into Lectern

### PR DESCRIPTION
**Pull request by Minecraftschurli**
this sadly needs an AT because the vanilla take book action happens in Screen/Container and not in the Tile

### CODE REVIEW

File: src/main/java/vazkii/patchouli/common/base/LecternEventHandler.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+import net.minecraftforge.eventbus.api.SubscribeEvent;
+import net.minecraftforge.fml.common.Mod;
+
+import vazkii.patchouli.api.PatchouliAPI;
+import vazkii.patchouli.common.book.Book;
+import vazkii.patchouli.common.item.ItemModBook;
+
+/**
+ * @author  Minecraftschurli
+ * @version 2020-05-01
```
Comment (williewillus) [inline src/main/java/vazkii/patchouli/common/base/LecternEventHandler.java:23]: remove this

Comment (Minecraftschurli) [reply src/main/java/vazkii/patchouli/common/base/LecternEventHandler.java:23]: ok will do

**Comment by Hubry** [comment]
This does not seem to take mods with their own book item implementations (like Botania) into account.

**Comment by Minecraftschurli** [comment]
Yes this is due to the difficulty to determine if it is a Patchouli book and if yes what the book id is to open the book

**Comment by Hubry** [comment]
Like this method that already exists?

https://github.com/Vazkii/Patchouli/blob/055a352dc38084d6cd70b7132ddf4af32b00c47b/src/main/java/vazkii/patchouli/client/handler/BookRightClickHandler.java#L91-L105

**Comment by Minecraftschurli** [comment]
Implemented it

**Comment by Extegral** [comment]
Perhaps I'm missing out on something, but I'm not sure I get the point of using access transformer in here. Why not just access inventory field with classical reflection?

**Comment by Minecraftschurli** [comment]
Because you should always prefer AT over reflection

**Comment by Extegral** [comment]
Why? I know some people prefer to always use Mixin for reflection purposes nowadays, for instance, but it's not like it's better, it's simply more convenient and simple.

**Comment by Minecraftschurli** [comment]
If you use Mixins for accessing a private field you're doing something wrong

**Comment by Extegral** [comment]
> If you use Mixins for accessing a private field you're doing something wrong

Okay, how about this https://github.com/Vazkii/Botania/blob/master/src/main/java/vazkii/botania/mixin/AccessorAbstractFurnaceTileEntity.java
...and a lot of other examples Botania has.

That also doesn't answer the question of why AT is supposedly better.

**Comment by Minecraftschurli** [comment]
@Extegral if you want to know why AT is better than mixin for making a field accessible ask in the forge discord

**Comment by Extegral** [comment]
@Minecraftschurli well, I know how both work from technical standpoint, so I doubt there's anything knew on this I would learn. Admittedly though since the time of my question I used AT's in actual implementations a bit more, and reasons why some would consider it a better solution are much more obvious now.

But on the funny side note, as @Mumfrey stated in [Release Notes Mixin 0.8#Notable Bug Fixes](https://github.com/SpongePowered/Mixin/wiki/Release-Notes---Mixin-0.8#notable-bug-fixes):

> It's my life mission to make _Access Transformers_ redundant

...which also makes a surprising lot of sense if you consider.

**Comment by Hubry** [comment]
How is this still not merged?

Anyways, please use accessor mixins over ATs, and rebase on master branch.

File: src/main/java/vazkii/patchouli/common/handler/LecternEventHandler.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+		BlockState state = world.getBlockState(pos);
+		if (!world.isRemote) {
+			TileEntity tileEntity = world.getTileEntity(pos);
+			if (tileEntity instanceof LecternTileEntity) {
+				PlayerEntity player = event.getPlayer();
+				if (state.get(LecternBlock.HAS_BOOK)) {
+					if (player.isSneaking()) {
+						takeBook(player, (LecternTileEntity) tileEntity);
+					} else {
+						Book book = BookRightClickHandler.getBookFromStack(((LecternTileEntity) tileEntity).getBook());
```
Comment (Hubry) [inline src/main/java/vazkii/patchouli/common/handler/LecternEventHandler.java:37]: Move this method into a common class because it's in a client package.

**Comment by Minecraftschurli** [comment]
Sorry but I never used mixins before, so I don't know how to use them. But I can do the rebase and the fix.

**Comment by Minecraftschurli** [comment]
DONE!
Removed AT and rebased

**Review by Hubry — APPROVED** [review]
I guess you could squash the commits, but this looks fine to me.

@williewillus please review lol

**Comment by Extegral** [comment]
> **Minecraftschurli**: If you use Mixins for accessing a private field you're doing something wrong

> **Hubry:** Anyways, please use accessor mixins over ATs, and rebase on master branch.

> **Minecraftschurli**: DONE!

Oh, look how the tables turned

**Comment by Hubry** [comment]
Haha, I am not too involved here (I got push access to skip the PR workflow where my PRs are pulled anyways, basically), but that collaborator badge is a bit convincing, isn't it :P 

ATs are very wasteful, even having a single one makes a project heavy because it now needs a local copy of the main Forge dep with the transforms applied. Both accessor mixins and reflection don't use that. And if you can skip any of those entirely...

**Comment by Minecraftschurli** [comment]
@Extegral well I don't if you look at my commits I'm using neither

**Comment by Extegral** [comment]
Oh, I guess this PR was stuck around for long enough for that `LecternBlock` to gain brand new `#setHasBook` method. How unfortunate now we have no means of ever truly knowing which method would've been preferred by the maintainers of this repository.

File: src/main/java/vazkii/patchouli/common/handler/LecternEventHandler.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+@Mod.EventBusSubscriber(bus = Mod.EventBusSubscriber.Bus.FORGE)
+public class LecternEventHandler {
+
+	@SubscribeEvent
+	static void rightClick(PlayerInteractEvent.RightClickBlock event) {
+		World world = event.getWorld();
+		BlockPos pos = event.getPos();
+		BlockState state = world.getBlockState(pos);
+		if (!world.isRemote) {
```
Comment (williewillus) [inline src/main/java/vazkii/patchouli/common/handler/LecternEventHandler.java:29]: push this check down. player interact events should ideally have the same results (event.setUseBlock/setUseItem) both client and serverside, or else you could get desyncs.

**Review by williewillus — CHANGES_REQUESTED** [review]
Looks fine other than the comment

File: src/main/java/vazkii/patchouli/common/handler/LecternEventHandler.java (hunk 3 above)
Comment (Minecraftschurli) [reply src/main/java/vazkii/patchouli/common/handler/LecternEventHandler.java]: done

**Pull request merged**